-strb	w8, [x24, #0x71e]
+strb	w8, [x25, #0x71e]
 mov	x0, x19
 mov	w1, #0x14
 bl	<<esp32c6::periph::Peripherals as esp_periph::mmio::Dispatch>::refresh_optional>
+ldrb	w8, [x19, #0xe69]
+tbnz	w8, #0x0, <<esp32c6::bus::SocBus>::periph_write>
+lsr	w8, w28, #12
+sub	w9, w8, #0x60, lsl #12
+cmp	w9, #0x15
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+mov	w9, #0x80
+movk	w9, #0x6, lsl #16
+cmp	w8, w9
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+add	w8, w22, #0x78
+cmp	w27, w8
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
 ldr	x4, [x19, #0x668]
 ldr	x5, [x19, #0x660]
-ldr	x8, [sp, #0x8]
-cmp	x8, x4
-b.ne	<<esp32c6::bus::SocBus>::periph_write>
-cmp	x23, x5
+cmp	x26, x4
+b.ne	<<esp32c6::bus::SocBus>::periph_write>
+cmp	x24, x5
 b.ne	<<esp32c6::bus::SocBus>::periph_write>
 mov	w8, #0x60000000
@@ -724,6 +768,7 @@

 b.ne	<<esp32c6::bus::SocBus>::periph_write>
 mov	w2, #0x1
-ldrb	w8, [x24, #0xe08]
+ldrb	w8, [x25, #0xe08]
 cbz	w8, <<esp32c6::bus::SocBus>::periph_write>
+mov	x23, x21
 ldr	x21, [x19, #0x2380]
 ldr	x20, [x19, #0x2388]
@@ -736,9 +781,9 @@

 mov	x0, x21
 mov	x1, x22
-mov	x3, x25
+mov	x3, x23
 blr	x8
-ldrb	w8, [x24, #0x7f8]
+ldrb	w8, [x25, #0x7f8]
 tbz	w8, #0x0, <<esp32c6::bus::SocBus>::periph_write>
-strb	wzr, [x24, #0x7f8]
+strb	wzr, [x25, #0x7f8]
 ldr	x1, [x19, #0x2350]
 ldr	x2, [x19, #0x2358]
@@ -749,10 +794,10 @@

 bl	<<esp_periph::spi_mem::SpiMem>::execute>
 str	xzr, [x19, #0xd20]
-ldrb	w8, [x24, #0x7f8]
+ldrb	w8, [x25, #0x7f8]
 tbnz	w8, #0x0, <<esp32c6::bus::SocBus>::periph_write>
 ldr	x8, [x19, #0x2158]
 ldrb	w8, [x8, #0x50]
 ubfx	w20, w8, #2, #1
-strb	w20, [x24, #0x7f9]
+strb	w20, [x25, #0x7f9]
 ldr	w8, [x19, #0x1e0]
 cbz	w8, <<esp32c6::bus::SocBus>::periph_write>
@@ -860,5 +905,5 @@

 bl	<<esp32c6::bus::SocBus>::deliver_board_events>
 mov	w8, #0x1
-strb	w8, [x24, #0xe09]
+strb	w8, [x25, #0xe09]
 ldp	x29, x30, [sp, #0x110]
 ldp	x20, x19, [sp, #0x100]
@@ -869,58 +914,4 @@

 add	sp, sp, #0x120
 ret
-ldr	x8, [x19, #0x2360]
-cmn	x8, #0x1
-b.eq	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x8, [x19, #0x658]
-cbz	x8, <<esp32c6::bus::SocBus>::periph_write>
-str	w27, [sp, #0x8]
-mov	x27, x25
-mov	x25, x24
-ldr	x21, [x19, #0x650]
-add	x22, x21, x8, lsl #1
-ldr	x23, [x19, #0x2370]
-lsl	x26, x23, #4
-b	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x8, [x19, #0x2368]
-add	x8, x8, x26
-str	x24, [x8]
-strb	w28, [x8, #0x8]
-strb	w20, [x8, #0x9]
-add	x23, x23, #0x1
-str	x23, [x19, #0x2370]
-add	x26, x26, #0x10
-add	x21, x21, #0x2
-cmp	x21, x22
-b.eq	<<esp32c6::bus::SocBus>::periph_write>
-ldrb	w28, [x21]
-ldrb	w20, [x21, #0x1]
-ldr	x24, [x19, #0x2790]
-ldr	x8, [x19, #0x2360]
-cmp	x23, x8
-b.ne	<<esp32c6::bus::SocBus>::periph_write>
-mov	w8, #0x2360
-add	x0, x19, x8
-bl	<<alloc::raw_vec::RawVec<(u64, u8, bool)>>::grow_one>
-b	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x4, [x19, #0x668]
-ldr	x5, [x19, #0x660]
-mov	x24, x25
-mov	x25, x27
-ldr	w27, [sp, #0x8]
-ldr	x0, [x19, #0x2380]
-ldr	x8, [x19, #0x2388]
-ldr	x1, [x19, #0x2790]
-ldr	x2, [x19, #0x650]
-ldr	x3, [x19, #0x658]
-ldr	x8, [x8, #0x28]
-blr	x8
-str	xzr, [x19, #0x658]
-mov	w8, #0x60000000
-cmp	w27, w8
-b.ne	<<esp32c6::bus::SocBus>::periph_write>
-mov	x2, #0x0
-ldrb	w8, [x24, #0xe08]
-cbnz	w8, <<esp32c6::bus::SocBus>::periph_write>
-b	<<esp32c6::bus::SocBus>::periph_write>
 cmp	w9, #0x2
 b.ne	<<esp32c6::bus::SocBus>::periph_write>
@@ -944,11 +935,12 @@

 tst	w8, #0x3
 b.eq	<<esp32c6::bus::SocBus>::periph_write>
+mov	x20, x24
 mov	w9, #0x0
 ldr	w8, [x19, #0x2038]
 ldr	w10, [x19, #0x203c]
-and	w21, w10, w8
+and	w23, w10, w8
 ldr	x8, [x19, #0x1f18]
-ldr	x22, [x19, #0x1f20]
-cbnz	x22, <<esp32c6::bus::SocBus>::periph_write>
+ldr	x24, [x19, #0x1f20]
+cbnz	x24, <<esp32c6::bus::SocBus>::periph_write>
 b	<<esp32c6::bus::SocBus>::periph_write>
 mov	w8, #0x0
@@ -956,5 +948,5 @@

 movk	w9, #0x262, lsl #16
 b	<<esp32c6::bus::SocBus>::periph_write>
-lsr	w0, w22, #2
+lsr	w0, w23, #2
 ldr	x1, [x19, #0x1d88]
 cmp	x1, x0
@@ -965,9 +957,11 @@

 mov	w8, #0x1f28
 add	x0, x19, x8
+stp	x12, x11, [sp]
 bl	<<alloc::raw_vec::RawVec<(u32, u32, u32, bool)>>::grow_one>
-b	<<esp32c6::bus::SocBus>::periph_write>
-cmp	w20, #0x180
-b.eq	<<esp32c6::bus::SocBus>::periph_write>
-dup.4s	v0, w20
+ldp	x12, x11, [sp]
+b	<<esp32c6::bus::SocBus>::periph_write>
+cmp	w11, #0x180
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+dup.4s	v0, w11
 adrp x8
 ldr	q1, [x8, #PAGEOFF]
@@ -980,31 +974,36 @@

 cbnz	x8, <<esp32c6::bus::SocBus>::periph_write>
 add	x0, x19, #0xb8
-and	w1, w21, #0xffc
+and	w1, w28, #0xffc
 mov	w2, #0x1
-mov	x3, x22
+mov	x3, x23
+mov	x20, x11
 bl	<<esp32c6::radio::Ieee802154>::note_unknown>
-b	<<esp32c6::bus::SocBus>::periph_write>
-str	w22, [x19, #0x1f0]
-b	<<esp32c6::bus::SocBus>::periph_write>
-str	w22, [x19, #0x200]
-b	<<esp32c6::bus::SocBus>::periph_write>
-str	w22, [x19, #0x2b4]
-b	<<esp32c6::bus::SocBus>::periph_write>
-str	w22, [x19, #0x2b8]
+lsr	w0, w20, #2
+ldr	x1, [x19, #0x180]
+cmp	x1, x0
+b.hi	<<esp32c6::bus::SocBus>::periph_write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+str	w23, [x19, #0x1f0]
+b	<<esp32c6::bus::SocBus>::periph_write>
+str	w23, [x19, #0x200]
+b	<<esp32c6::bus::SocBus>::periph_write>
+str	w23, [x19, #0x2b4]
+b	<<esp32c6::bus::SocBus>::periph_write>
+str	w23, [x19, #0x2b8]
 b	<<esp32c6::bus::SocBus>::periph_write>
 ldr	w8, [x19, #0x2a0]
-bic	w8, w8, w22
+bic	w8, w8, w23
 str	w8, [x19, #0x2a0]
 b	<<esp32c6::bus::SocBus>::periph_write>
-and	w8, w22, #0x7f
+and	w8, w23, #0x7f
 str	w8, [x19, #0x2a8]
 b	<<esp32c6::bus::SocBus>::periph_write>
-and	w8, w22, #0x1fff
+and	w8, w23, #0x1fff
 str	w8, [x19, #0x2a4]
 b	<<esp32c6::bus::SocBus>::periph_write>
-and	w8, w22, #0xffffff
+and	w8, w23, #0xffffff
 str	w8, [x19, #0x2b0]
 b	<<esp32c6::bus::SocBus>::periph_write>
-and	w8, w22, #0x1f
+and	w8, w23, #0x1f
 str	w8, [x19, #0x2ac]
 b	<<esp32c6::bus::SocBus>::periph_write>
@@ -1013,8 +1012,8 @@

 b.ls	<<esp32c6::bus::SocBus>::periph_write>
 ldr	x8, [x19, #0x178]
-and	w9, w22, #0xffff
+and	w9, w23, #0xffff
 str	w9, [x8, #0x54]
 b	<<esp32c6::bus::SocBus>::periph_write>
-and	w8, w22, #0xff
+and	w8, w23, #0xff
 stur	w8, [x29, #-0x70]
 ldrb	w9, [x19, #0x2cd]
@@ -1045,55 +1044,58 @@

 strb	wzr, [x19, #0x204]
 b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0xf00
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::rsa::Rsa as esp_periph::device::Device>::write>
+mov	w8, #0x1f10
+add	x0, x19, x8
+bl	<<alloc::raw_vec::RawVec<u32>>::grow_one>
+ldr	x8, [x19, #0x1f18]
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0xd80
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::sha::Sha as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+lsr	w0, w11, #2
+ldr	x1, [x19, #0xc50]
+cmp	x1, x0
+b.ls	<<esp32c6::bus::SocBus>::periph_write>
+ldr	x8, [x19, #0xc48]
+str	w5, [x8, x0, lsl #2]
+b	<<esp32c6::bus::SocBus>::periph_write>
+mov	w8, #0x2048
+add	x0, x19, x8
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::mcpwm::Mcpwm as esp_periph::device::Device>::write>
+mov	x0, x19
+mov	w1, #0x14
+bl	<<esp32c6::periph::Peripherals as esp_periph::mmio::Dispatch>::refresh_optional>
+b	<<esp32c6::bus::SocBus>::periph_write>
+mov	w8, #0x19b8
+add	x0, x19, x8
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp32c6::wifi::WifiMac as esp_periph::device::Device>::write>
 b	<<esp32c6::bus::SocBus>::periph_write>
 ldr	x1, [x19, #0x1e88]
-lsr	w0, w20, #2
+lsr	w0, w11, #2
 cmp	x1, x0
 b.ls	<<esp32c6::bus::SocBus>::periph_write>
 ldr	x8, [x19, #0x1e80]
 str	w5, [x8, x0, lsl #2]
-b	<<esp32c6::bus::SocBus>::periph_write>
-mov	w8, #0x2048
-add	x0, x19, x8
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::mcpwm::Mcpwm as esp_periph::device::Device>::write>
-mov	x0, x19
-mov	w1, #0x14
-bl	<<esp32c6::periph::Peripherals as esp_periph::mmio::Dispatch>::refresh_optional>
-b	<<esp32c6::bus::SocBus>::periph_write>
-and	w8, w21, #0xc00
-cmp	w8, #0x400
-b.eq	<<esp32c6::bus::SocBus>::periph_write>
-cmp	w8, #0x800
-b.ne	<<esp32c6::bus::SocBus>::periph_write>
-sub	w8, w20, #0x800
-cmp	w8, #0x1d4
-b.eq	<<esp32c6::bus::SocBus>::periph_write>
-lsr	w0, w8, #2
-ldr	x1, [x19, #0xc68]
-cmp	x1, x0
-b.ls	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x8, [x19, #0xc60]
-str	w5, [x8, x0, lsl #2]
-b	<<esp32c6::bus::SocBus>::periph_write>
-mov	w8, #0x19b8
-add	x0, x19, x8
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp32c6::wifi::WifiMac as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-mov	x20, x23
+tst	w5, #0x2
+mov	w8, #0xd8
+ccmp	w11, w8, #0x0, ne
+mov	w8, #0x8
+csel	w23, wzr, w8, ne
+b	<<esp32c6::bus::SocBus>::periph_write>
+mov	x20, x24
 ldr	x0, [x19, #0x2158]
-and	w1, w21, #0xffc
+mov	x22, x28
+and	w1, w28, #0xffc
 mov	x2, x5
 bl	<<esp_periph::i2s::I2s as esp_periph::device::Device>::write>
-and	w22, w0, #0xff
+and	w23, w0, #0xff
 ldr	x8, [x19, #0x2158]
 ldp	w9, w8, [x8, #0x58]
-and	w23, w8, w9
+and	w24, w8, w9
 ldr	x28, [x19, #0x1f20]
 cbz	x28, <<esp32c6::bus::SocBus>::periph_write>
@@ -1108,14 +1110,56 @@

 b.ne	<<esp32c6::bus::SocBus>::periph_write>
 b	<<esp32c6::bus::SocBus>::periph_write>
-cmp	w20, #0x8b
-b.gt	<<esp32c6::bus::SocBus>::periph_write>
-cbz	w20, <<esp32c6::bus::SocBus>::periph_write>
-cmp	w20, #0x4
-b.ne	<<esp32c6::bus::SocBus>::periph_write>
-str	w5, [x19, #0x1e14]
+subs	w1, w11, #0x400
+b.hs	<<esp32c6::bus::SocBus>::periph_write>
+mov	w8, #0x1e48
+add	x0, x19, x8
+orr	w1, w11, #0x400
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0x618
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::gpio::Gpio as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+lsr	w0, w11, #2
+cmp	w11, #0x134
+b.hs	<<esp32c6::bus::SocBus>::periph_write>
+add	x8, x19, x0, lsl #2
+and	w9, w5, #0x1f
+str	w9, [x8, #0x1c40]
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0xf00
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::rsa::Rsa as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0xc70
+b	<<esp32c6::bus::SocBus>::periph_write>
+ldr	x1, [x19, #0x1ee8]
+lsr	w0, w11, #2
+cmp	x1, x0
+b.ls	<<esp32c6::bus::SocBus>::periph_write>
+ldr	x8, [x19, #0x1ee0]
+str	w5, [x8, x0, lsl #2]
+b	<<esp32c6::bus::SocBus>::periph_write>
+ldr	x1, [x19, #0x1e40]
+lsr	w0, w11, #2
+cmp	x1, x0
+b.ls	<<esp32c6::bus::SocBus>::periph_write>
+ldr	x8, [x19, #0x1e38]
+str	w5, [x8, x0, lsl #2]
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0xf58
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::rmt_compact::RmtCompact as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0x468
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::usb_serial_jtag::UsbSerialJtag as esp_periph::device::Device>::write>
 b	<<esp32c6::bus::SocBus>::periph_write>
 ubfx	w8, w5, #14, #1
 ldr	x1, [x19, #0x19a0]
-cmp	w20, #0xc0
+cmp	w11, #0xc0
 csel	w8, wzr, w8, ne
 tbz	w8, #0x0, <<esp32c6::bus::SocBus>::periph_write>
@@ -1131,6 +1175,6 @@

 str	w8, [x19, #0x19ac]
 mov	w8, #0x1
-strb	w8, [x24]
-lsr	w0, w20, #2
+strb	w8, [x25]
+lsr	w0, w11, #2
 cmp	x1, x0
 b.ls	<<esp32c6::bus::SocBus>::periph_write>
@@ -1138,53 +1182,13 @@

 str	w5, [x8, x0, lsl #2]
 b	<<esp32c6::bus::SocBus>::periph_write>
-lsr	w0, w20, #2
-cmp	w20, #0x134
-b.hs	<<esp32c6::bus::SocBus>::periph_write>
-add	x8, x19, x0, lsl #2
-and	w9, w5, #0x1f
-str	w9, [x8, #0x1c40]
-b	<<esp32c6::bus::SocBus>::periph_write>
-lsr	w0, w20, #2
-ldr	x1, [x19, #0xc50]
-cmp	x1, x0
-b.ls	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x8, [x19, #0xc48]
-str	w5, [x8, x0, lsl #2]
-b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0xc70
-b	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x1, [x19, #0x1e40]
-lsr	w0, w20, #2
-cmp	x1, x0
-b.ls	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x8, [x19, #0x1e38]
-str	w5, [x8, x0, lsl #2]
-b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0xf58
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::rmt_compact::RmtCompact as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0x468
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::usb_serial_jtag::UsbSerialJtag as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-subs	w1, w20, #0x800
-b.lo	<<esp32c6::bus::SocBus>::periph_write>
-mov	w8, #0x1e90
-add	x0, x19, x8
-mov	x2, x5
-bl	<<esp32c6::periph::AnaMst as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
 add	x0, x19, #0x400
-and	w1, w21, #0xffc
+and	w1, w28, #0xffc
 mov	x2, x5
 bl	<<esp_periph::uart::Uart as esp_periph::device::Device>::write>
 b	<<esp32c6::bus::SocBus>::periph_write>
-mov	x26, x23
+str	x24, [sp, #0x8]
 mov	w20, #0x1fb8
 add	x0, x19, x20
-and	w1, w21, #0xffc
+and	w1, w28, #0xffc
 mov	x2, x5
 bl	<<esp_periph::ledc::Ledc as esp_periph::device::Device>::write>
@@ -1193,10 +1197,10 @@

 ldr	w8, [x19, #0x2038]
 ldr	w9, [x19, #0x203c]
-and	w22, w9, w8
+and	w23, w9, w8
 ldr	x8, [x19, #0x1f18]
-ldr	x23, [x19, #0x1f20]
-cbz	x23, <<esp32c6::bus::SocBus>::periph_write>
+ldr	x24, [x19, #0x1f20]
+cbz	x24, <<esp32c6::bus::SocBus>::periph_write>
 mov	x9, #0x0
-lsl	x10, x23, #2
+lsl	x10, x24, #2
 ldr	w11, [x8, x9, lsl #2]
 cmp	w11, #0x7
@@ -1209,86 +1213,118 @@

 b.eq	<<esp32c6::bus::SocBus>::periph_write>
 ldr	x9, [x19, #0x1f10]
-cmp	x23, x9
+cmp	x24, x9
 b.eq	<<esp32c6::bus::SocBus>::periph_write>
 mov	w9, #0x7
-str	w9, [x8, x23, lsl #2]
+str	w9, [x8, x24, lsl #2]
 mov	w8, #0x1
 b	<<esp32c6::bus::SocBus>::periph_write>
 add	x0, x19, #0x518
+b	<<esp32c6::bus::SocBus>::periph_write>
+ldr	x8, [x19, #0x2158]
+add	x0, x8, #0xf0
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp32c6::parlio::Parlio as esp_periph::device::Device>::write>
+ldr	x8, [x19, #0x2158]
+ldr	x8, [x8, #0xf0]
+cbz	x8, <<esp32c6::bus::SocBus>::periph_write>
+mov	x20, x24
+ldp	w9, w8, [x8, #0x74]
+tst	w8, w9
+mov	w8, #-0x80000000
+csel	w23, wzr, w8, eq
 b	<<esp32c6::bus::SocBus>::periph_write>
 mov	w8, #0x16e0
 add	x0, x19, x8
-and	w1, w21, #0xffc
+and	w1, w28, #0xffc
 mov	x2, x5
 bl	<<esp32c6::periph::GdmaC6 as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+cmp	w11, #0x8b
+b.gt	<<esp32c6::bus::SocBus>::periph_write>
+cbz	w11, <<esp32c6::bus::SocBus>::periph_write>
+cmp	w11, #0x4
+b.ne	<<esp32c6::bus::SocBus>::periph_write>
+str	w5, [x19, #0x1e14]
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0xcf8
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::spi_mem::SpiMem as esp_periph::device::Device>::write>
+and	w23, w0, #0xff
+b	<<esp32c6::bus::SocBus>::periph_write>
+and	w1, w28, #0xffc
+mov	x0, x19
+mov	x2, x5
+bl	<<esp_periph::i2c::I2c as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+mov	w8, #0x20d0
+add	x0, x19, x8
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::systimer::Systimer as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0x598
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::timg::TimerGroup as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+add	x0, x19, #0xe70
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp_periph::aes::Aes as esp_periph::device::Device>::write>
 b	<<esp32c6::bus::SocBus>::periph_write>
 mov	w8, #0x18e8
 add	x0, x19, x8
-and	w1, w21, #0xffc
+and	w1, w28, #0xffc
 mov	x2, x5
 bl	<<esp_periph::gpspi::GpSpi as esp_periph::device::Device>::write>
 b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0xcf8
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::spi_mem::SpiMem as esp_periph::device::Device>::write>
-and	w22, w0, #0xff
-b	<<esp32c6::bus::SocBus>::periph_write>
-and	w1, w21, #0xffc
-mov	x0, x19
-mov	x2, x5
-bl	<<esp_periph::i2c::I2c as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-mov	w8, #0x20d0
-add	x0, x19, x8
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::systimer::Systimer as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0x598
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::timg::TimerGroup as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0xd80
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::sha::Sha as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0xe70
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::aes::Aes as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-add	x0, x19, #0x618
-and	w1, w21, #0xffc
-mov	x2, x5
-bl	<<esp_periph::gpio::Gpio as esp_periph::device::Device>::write>
-b	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x1, [x19, #0x1ee8]
-lsr	w0, w20, #2
+mov	x20, x24
+ldr	x8, [x19, #0x2158]
+add	x0, x8, #0x80
+and	w1, w28, #0xffc
+mov	x2, x5
+bl	<<esp32c6::ecc::Ecc as esp_periph::device::Device>::write>
+ldr	x8, [x19, #0x2158]
+ldp	w9, w8, [x8, #0xe4]
+and	w23, w8, w9
+ldr	x24, [x19, #0x1f20]
+cbz	x24, <<esp32c6::bus::SocBus>::periph_write>
+mov	x8, #0x0
+ldr	x9, [x19, #0x1f18]
+lsl	x10, x24, #2
+ldr	w11, [x9, x8, lsl #2]
+cmp	w11, #0x8b
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+add	x8, x8, #0x1
+subs	x10, x10, #0x4
+b.ne	<<esp32c6::bus::SocBus>::periph_write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+and	w8, w28, #0xc00
+cmp	w8, #0x400
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+cmp	w8, #0x800
+b.ne	<<esp32c6::bus::SocBus>::periph_write>
+sub	w8, w11, #0x800
+cmp	w8, #0x1d4
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+lsr	w0, w8, #2
+ldr	x1, [x19, #0xc68]
 cmp	x1, x0
 b.ls	<<esp32c6::bus::SocBus>::periph_write>
-ldr	x8, [x19, #0x1ee0]
+ldr	x8, [x19, #0xc60]
 str	w5, [x8, x0, lsl #2]
 b	<<esp32c6::bus::SocBus>::periph_write>
-subs	w1, w20, #0x400
-b.hs	<<esp32c6::bus::SocBus>::periph_write>
+subs	w1, w11, #0x800
+b.lo	<<esp32c6::bus::SocBus>::periph_write>
+mov	w8, #0x1e90
+add	x0, x19, x8
+mov	x2, x5
+bl	<<esp32c6::periph::AnaMst as esp_periph::device::Device>::write>
+b	<<esp32c6::bus::SocBus>::periph_write>
+cmp	w11, #0xc00
+b.lo	<<esp32c6::bus::SocBus>::periph_write>
 mov	w8, #0x1e48
-add	x0, x19, x8
-orr	w1, w20, #0x400
-b	<<esp32c6::bus::SocBus>::periph_write>
-mov	w8, #0x1f10
-add	x0, x19, x8
-bl	<<alloc::raw_vec::RawVec<u32>>::grow_one>
-ldr	x8, [x19, #0x1f18]
-b	<<esp32c6::bus::SocBus>::periph_write>
-cmp	w20, #0x8c
-b.eq	<<esp32c6::bus::SocBus>::periph_write>
-cmp	w20, #0xa8
-b.ne	<<esp32c6::bus::SocBus>::periph_write>
-ldr	w8, [x19, #0x1e1c]
-bic	w8, w8, w5
-str	w8, [x19, #0x1e1c]
 b	<<esp32c6::bus::SocBus>::periph_write>
 ldr	x1, [x19, #0x1c38]
@@ -1298,7 +1334,11 @@

 str	w5, [x8, x0, lsl #2]
 b	<<esp32c6::bus::SocBus>::periph_write>
-cmp	w20, #0xc00
-b.lo	<<esp32c6::bus::SocBus>::periph_write>
-mov	w8, #0x1e48
+cmp	w11, #0x8c
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+cmp	w11, #0xa8
+b.ne	<<esp32c6::bus::SocBus>::periph_write>
+ldr	w8, [x19, #0x1e1c]
+bic	w8, w8, w5
+str	w8, [x19, #0x1e1c]
 b	<<esp32c6::bus::SocBus>::periph_write>
 cmp	w8, #0x42
@@ -1331,13 +1371,27 @@

 strb	w8, [x19, #0x2cf]
 b	<<esp32c6::bus::SocBus>::periph_write>
+mov	x20, x24
+mov	w23, #0x0
+ldr	x24, [x19, #0x1f20]
+cbz	x24, <<esp32c6::bus::SocBus>::periph_write>
+mov	x8, #0x0
+ldr	x9, [x19, #0x1f18]
+lsl	x10, x24, #2
+ldr	w11, [x9, x8, lsl #2]
+cmp	w11, #0x15
+b.eq	<<esp32c6::bus::SocBus>::periph_write>
+add	x8, x8, #0x1
+subs	x10, x10, #0x4
+b.ne	<<esp32c6::bus::SocBus>::periph_write>
+b	<<esp32c6::bus::SocBus>::periph_write>
 mov	w8, #0x1e48
-sub	w1, w20, #0x400
+sub	w1, w11, #0x400
 add	x0, x19, x8
 b	<<esp32c6::bus::SocBus>::periph_write>
-cmp	w20, #0xc00
+cmp	w11, #0xc00
 b.lo	<<esp32c6::bus::SocBus>::periph_write>
 mov	w8, #0x1e48
 add	x0, x19, x8
-and	w1, w21, #0xffc
+and	w1, w28, #0xffc
